fuse_loss/test/test_huber_loss.cpp

```diff
@@ -0,0 +1,162 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2018, Locus Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_core/serialization.h>
+#include <fuse_loss/huber_loss.h>
+
+#include <ceres/autodiff_cost_function.h>
+#include <ceres/problem.h>
+#include <ceres/solver.h>
+#include <gtest/gtest.h>
+
+TEST(HuberLoss, Constructor)
```

Comment:
I just realized I only create a unit test for the `HuberLoss`, and not for the others. Honestly, the unit test for the others would be almost the same, so I'm not sure if it's worth having it or not. If you think it's good idea or you don't have a strong opinion, I can create them. I'd basically copy this and create the others changing the loss function used and updating the parameters they take, as well as the values, so the optimization takes makes sense.

These classes aren't really giving a lot more functionality than the `ceres::LossFunction`s, so I probably shouldn't be testing much here. But I still thing the optimization test below is a good thing to have, either on this pkg or in another one inside the `fuse` repo.

Thoughts?

Reply:
Given the "thin wrapper"-ness of the `fuse_core::Loss` class, testing just one of the classes seems reasonable. I would add an additional check to the serialization unit to verify the `ceres::LossFunction*` survives the serialize-deserialize process. As mentioned earlier, I don't think it currently does.

And the only other unit test I would like to see is an optimization of a graph where the final value is different when the loss function is used, versus when it is not. That would verify that the loss function is correctly making it all the way into the Ceres problem. That would belong in the fuse_graph package, I suppose. Here we are manually creating the problem, so we are not testing the real code path used by the optimizers.

Reply:
You were right. I just extended the test to check the `ceres::LossFunction*` survices the serialize-deserialize process.

I check:
1. It's not NULL, which it's not enough because it could get constructed again with default values, instead of the originals
2. It satisfies some invariantes for the inlier region
3. It satisfies some invariantes for the outlier region

There's already a unit test I added that exercise and checks the error changes depending on whether the loss functions are used or not. That test is actually in `fuse_loss`. Not sure if you want me to move it somewhere else. :thinking: Note that I updated some existing tests in both `fuse_constraints` and `fuse_graphs`. I think they already test what you're saying.

Reply:
This looks good. Thank you.